rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs

```diff
@@ -97,90 +97,269 @@ impl From<(u64, Duration)> for Throughput {
+    // can differentiate between `Empty` due to there not having been enough
+    // time to establish a full buffer worth of data vs. `Empty` due to a
+    // polling gap. Once the length reaches N, it will never change again.
+    length: usize,
+}
+impl<const N: usize> LogBuffer<N> {
+    fn new() -> Self {
+        Self {
+            entries: [Bin::Empty; N],
+            length: 0,
+        }
+    }
+
+    /// Mutably returns the tail of the buffer.
+    ///
+    /// ## Panics
+    ///
+    /// The buffer MUST have at least one bin it it before this is called.
+    fn tail_mut(&mut self) -> &mut Bin {
+        debug_assert!(self.length > 0);
+        &mut self.entries[self.length - 1]
+    }
+
+    /// Pushes a bin into the buffer. If the buffer is already full,
+    /// then this will rotate the entire buffer to the left.
+    fn push(&mut self, bin: Bin) {
+        if self.filled() {
+            self.entries.rotate_left(1);
+            self.entries[N - 1] = bin;
+        } else {
+            self.entries[self.length] = bin;
+            self.length += 1;
+        }
+    }
+
+    /// Returns the total number of bytes transferred within the time window.
+    fn bytes_transferred(&self) -> u64 {
+        self.entries.iter().take(self.length).map(Bin::bytes).sum()
+    }
+
+    #[inline]
+    fn filled(&self) -> bool {
+        self.length == N
+    }
+
+    /// Fills in missing NoData entries.
+    ///
+    /// We want NoData entries to represent when a future hasn't been polled.
+    /// Since the future is in charge of logging in the first place, the only
+    /// way we can know about these is by examining gaps in time.
+    fn fill_gaps(&mut self) {
+        for entry in self.entries.iter_mut().take(self.length) {
+            if matches!(entry, Bin::Empty) {
+                *entry = Bin::NoPolling;
+            }
+        }
+    }
+
+    /// Returns the counts of each bin type in the buffer.
+    fn counts(&self) -> BinCounts {
+        let mut counts = BinCounts::default();
+        for entry in &self.entries {
+            match entry {
+                Bin::Empty => counts.empty += 1,
+                Bin::NoPolling => counts.no_polling += 1,
+                Bin::TransferredBytes(_) => counts.transferred += 1,
+                Bin::Pending(_) => counts.pending += 1,
+            }
+        }
+        counts
+    }
+}
+
+/// Report/summary of all the events in a time window.
+#[cfg_attr(test, derive(Debug, Eq, PartialEq))]
+pub(crate) enum ThroughputReport {
+    /// Not enough data to draw any conclusions. This happens early in a request/response.
+    Incomplete,
+    /// The stream hasn't been polled for most of this time window.
+    NoPolling,
+    /// The stream has been waiting for most of the time window.
+    Pending,
+    /// The stream transferred this amount of throughput during the time window.
+    Transferred(Throughput),
+}
+
+const BUFFER_SIZE: usize = 10;
```

Comment:
this is the number of throughput windows?

Reply:
Should rename this to `BIN_COUNT`.